zookeeper-specifications/zab-1.0/ZabWithFLE.tla

```diff
@@ -0,0 +1,1020 @@
+   and sends back ACK-LD to leader. *)
+FollowerHandleNEWLEADER(i, j) ==
+        /\ state[i] = FOLLOWING
+        /\ msgs[j][i] /= << >>
+        /\ msgs[j][i][1].mtype = NEWLEADER
+        /\ LET msg     == msgs[j][i][1]
+               infoOk  == /\ leaderAddr[i] = j
+                          /\ acceptedEpoch[i] = msg.mepoch
+               correct == /\ infoOk
+                          /\ zabState[i] = SYNCHRONIZATION
+                          /\ synced[i]
+                          /\ ZxidEqual(lastZxid[i], msg.mlastZxid)
+           IN /\ infoOk
+              /\ currentEpoch' = [currentEpoch EXCEPT ![i] = msg.mepoch]
+              /\ UpdateProposal(i, j, lastZxid[i], currentEpoch'[i])
+              /\ \/ /\ correct
+                    /\ Reply(i, j, [mtype  |-> ACKLD,
+                                    mepoch |-> msg.mepoch])
+                    /\ UNCHANGED inherentViolated
+                 \/ /\ ~correct
+                    /\ PrintT("Exception: Condition correct is false in FollowerHandleNEWLEADER(" \o ToString(i) \o ", " \o ToString(j) \o ").")
+                    /\ inherentViolated' = TRUE
+                    /\ Discard(j, i)
+        /\ UNCHANGED <<state, lastZxid, zabState, acceptedEpoch, history, commitIndex, logicalClock, receiveVotes, outOfElection, recvQueue, waitNotmsg,
+                       leaderVarsZ, tempVarsZ, followerVarsZ, proposalMsgsLog, epochLeader, electionMsgs, idTable>>
+
+\* In phase l22, leader receives ACK-LD from a quorum of followers, and sends COMMIT-LD(UPTODATE) to followers.
+LeaderHandleACKLD(i, j) ==
+        /\ state[i] = LEADING
+        /\ msgs[j][i] /= << >>
+        /\ msgs[j][i][1].mtype = ACKLD
+        /\ LET msg     == msgs[j][i][1]
+               infoOk  == /\ acceptedEpoch[i] = msg.mepoch
+                          /\ j \in learners[i]
+               replyOk == /\ infoOk
+                          /\ NullPoint \in ackldRecv[i]
+           IN /\ infoOk
+              /\ \/ /\ replyOk
+                    /\ Reply(i, j, [mtype   |-> UPTODATE,
+                                    mepoch  |-> acceptedEpoch[i],
+                                    mcommit |-> commitIndex[i]])
+                    /\ committedCounter' = [committedCounter EXCEPT ![i][j] = Maximum({commitIndex[i] - Len(initialHistory[i]), committedCounter[i][j]})]
+                 \/ /\ ~replyOk
+                    /\ Discard(j, i)
+                    /\ UNCHANGED committedCounter
+              /\ ackldRecv' = [ackldRecv EXCEPT ![i] = IF j \notin ackldRecv[i] THEN ackldRecv[i] \union {j}
+                                                                                ELSE ackldRecv[i]]
+        /\ UNCHANGED <<serverVarsZ, electionVarsZ, leadingVoteSet, learners, cepochRecv, ackeRecv, forwarding, 
+                       ackIndex, currentCounter, sendCounter, committedIndex, tempVarsZ, followerVarsZ, verifyVarsZ, electionMsgs, idTable>>
+
+LeaderSync2(i) ==
```

Comment:
Action name is not descriptive, perhaps rename to something like LeaderTransitionsToBroadcast